src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs

```diff
@@ -2656,22 +2656,27 @@ private string GetCustomMarkupExtensionValue(XamlMemberDefinition member)
 
 			// Get the attribute from the custom markup extension class then get the return type specifed with MarkupExtensionReturnTypeAttribute
 			var attributeData = markupType.FindAttribute(XamlConstants.Types.MarkupExtensionReturnTypeAttribute);
+			var returnType = attributeData?.NamedArguments.FirstOrDefault(kvp => kvp.Key == "ReturnType").Value.Value;
 
-			if (attributeData == null)
-			{
-				this.Log().Error($"The custom markup extension {markupTypeDef.Type.Name} must specify the return type using {nameof(XamlConstants.Types.MarkupExtensionReturnTypeAttribute)}.");
-				return string.Empty;
-			}
+			var provideValue = $"(({xamlMarkupFullName})(new {markupTypeFullName} {{ {properties} }})).ProvideValue()";
 
-			var returnType = attributeData.NamedArguments.FirstOrDefault(kvp => kvp.Key == "ReturnType").Value.Value;
+			// If we cannot determine the return type needed to cast against the markup extension,
+			// we must then try to get the type of the target property, and use Convert.ChangeType()
+			if (returnType == null && FindPropertyType(member.Member) is INamedTypeSymbol propertyType)
+			{
+				var targetTypeDisplay = propertyType.ToDisplayString();
+				var targetType = $"typeof({targetTypeDisplay})";
 
-			if (returnType == null)
+				// It's important to cast to string before performing the conversion
+				return $"({targetTypeDisplay})(Convert.ChangeType(({provideValue}).ToString(), {targetType}))";
```

Comment:
`ChangeType` is probably only correct for supported target types (double, int etc)

Test with the following:
`ChangeType` is probably only correct for supported target types (double, int etc)

Test with the following:
```csharp
	public class MyMarkupExtension : MarkupExtension
	{
		public object Value1 { get; set; }
		protected override object ProvideValue()
		{
			return Value1;
		}
	}

        public class MyPoco
        {
            public double Bogosity { get; set; }
        }

    public partial class MyControl : Control
    {
        public MyPoco Poco
        {
            get { return (MyPoco)GetValue(PocoProperty); }
            set { SetValue(PocoProperty, value); }
        }
		
        public static readonly DependencyProperty PocoProperty =
            DependencyProperty.Register("Poco", typeof(MyPoco), typeof(MyControl), new PropertyMetadata(null, OnPocoChanged));

        private static void OnPocoChanged(DependencyObject dependencyObject, DependencyPropertyChangedEventArgs args)
        {
            ;
        }



		public object MyUntypedProperty
		{
			get { return (object)GetValue(MyUntypedPropertyProperty); }
			set { SetValue(MyUntypedPropertyProperty, value); }
		}
		
		public static readonly DependencyProperty MyUntypedPropertyProperty =
			DependencyProperty.Register("MyUntypedProperty", typeof(object), typeof(MyControl), new PropertyMetadata(null, OnMyUntypedPropertyChanged));

		private static void OnMyUntypedPropertyChanged(DependencyObject dependencyObject, DependencyPropertyChangedEventArgs args)
		{
			;
		}
```

```xaml
    <Page.Resources>
        <local:MyPoco x:Key="LocalPoco" Bogosity="27" />
    </Page.Resources>

    ...

            <local:MyControl MyUntypedProperty="{local:MyMarkup Value1={StaticResource LocalPoco}}" />
            <local:MyControl Poco="{local:MyMarkup Value1={StaticResource LocalPoco}}" />

```

Reply:
Done 👍  unit tests and source generation tests added + changes made